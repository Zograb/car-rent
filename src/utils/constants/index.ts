export * from './enums';
export * from './tailwindConfig';
