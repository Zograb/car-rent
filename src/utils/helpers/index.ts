export * from './formatNumberToCurrency';
export * from './getPercentageDifference';
export * from './getRandomColor';
export * from './getCurrentBreakpoint';
