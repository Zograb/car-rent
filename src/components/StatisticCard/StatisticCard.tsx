import { useMemo, type FC } from 'react';
import { AiOutlineArrowDown } from '@react-icons/all-files/ai/AiOutlineArrowDown';
import { AiOutlineArrowUp } from '@react-icons/all-files/ai/AiOutlineArrowUp';
import tw from 'twin.macro';

// Utils
import { formatNumberToCurrency, getPercentageDifference } from '@/utils';

interface StatisticCardProps {
  // card title
  title: string;

  // today revenue
  todayValue: number;

  // yesterday revenue
  yesterdayValue: number;

  // last week revenue
  lastWeekValue: number;
}

const StatisticCard: FC<StatisticCardProps> = ({
  title,
  todayValue,
  yesterdayValue,
  lastWeekValue,
}) => {
  const percentageDifference: number = useMemo(() => {
    return getPercentageDifference(todayValue, yesterdayValue);
  }, [todayValue, yesterdayValue]);

  return (
    <div tw="bg-white p-6 rounded-lg shadow-md" data-testid="statistic-card">
      <div css={[
        tw`flex justify-between items-center`,
        tw`border-b border-solid border-gray-200`,
        tw`pb-3 text-gray3`,
      ]}>
        <div tw="font-medium text-lg">{title}</div>
        <div tw="bg-gray-100 py-1 px-2 rounded-sm text-xs">Today</div>
      </div>
      <div tw="flex justify-between items-end mt-3 mb-2">
        <div tw="text-black text-3xl font-bold">
          {formatNumberToCurrency(todayValue)}
        </div>
        <div
          css={[
            tw`flex justify-end items-center`,
            percentageDifference < 0 && tw`text-secondary`,
            percentageDifference > 0 && tw`text-green`,
          ]}
        >
          {percentageDifference < 0 && <AiOutlineArrowDown data-testid="arrow-down" />}
          {percentageDifference > 0 && <AiOutlineArrowUp data-testid="arrow-up" />}
          <span>{Math.abs(percentageDifference)}%</span>
        </div>
      </div>
      <div tw="text-sm text-gray3 font-light">
        Compared to ${yesterdayValue} yesterday
      </div>
      <div tw="text-sm text-gray2 font-medium mt-0.5 flex justify-between items-center">
        <div>Last week Income</div>
        <div>{formatNumberToCurrency(lastWeekValue)}</div>
      </div>
    </div>
  );
}

export default StatisticCard;
