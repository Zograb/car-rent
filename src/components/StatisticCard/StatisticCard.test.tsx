import { render, screen } from '@testing-library/react';

// Components
import StatisticCard from './StatisticCard';

// Utils
import { formatNumberToCurrency } from '@/utils';

describe('StatisticCard', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <StatisticCard
        title="Title"
        todayValue={9000}
        yesterdayValue={12000}
        lastWeekValue={50000}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render statistic card', () => {
    const todayValue = 10000;
    render(
      <StatisticCard
        title="Title"
        todayValue={todayValue}
        yesterdayValue={12000}
        lastWeekValue={50000}
      />
    );
    const statisticCard = screen.getByTestId('statistic-card');
    const statisticCardTitle = screen.getByText('Title');
    const statisticCardTodayValue = screen.getByText(
      formatNumberToCurrency(todayValue),
    );
    expect(statisticCard).toBeInTheDocument();
    expect(statisticCardTitle).toBeInTheDocument();
    expect(statisticCardTodayValue).toBeInTheDocument();
  });

  it('should render statistic card with down arrow if today value less then yesterday value', () => {
    render(
      <StatisticCard
        title="Title"
        todayValue={9000}
        yesterdayValue={12000}
        lastWeekValue={50000}
      />
    );
    const statisticCardArrow = screen.getByTestId('arrow-down');
    expect(statisticCardArrow).toBeInTheDocument();
  });

  it('should render statistic card with down arrow if today value more then yesterday value', () => {
    render(
      <StatisticCard
        title="Title"
        todayValue={14000}
        yesterdayValue={12000}
        lastWeekValue={50000}
      />
    );
    const statisticCardArrow = screen.getByTestId('arrow-up');
    expect(statisticCardArrow).toBeInTheDocument();
  });
});

