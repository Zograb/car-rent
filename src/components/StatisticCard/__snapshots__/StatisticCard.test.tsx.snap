// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticCard should match snapshot 1`] = `
.c0 {
  border-radius: 0.5rem;
  --tw-bg-opacity: 1;
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  padding: 1.5rem;
  --tw-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1),0 2px 4px -2px rgb(0 0 0 / 0.1);
  --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color),0 2px 4px -2px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
}

.c1 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-bottom-width: 1px;
  border-style: solid;
  --tw-border-opacity: 1;
  border-color: rgb(229 231 235 / var(--tw-border-opacity));
  padding-bottom: 0.75rem;
  --tw-text-opacity: 1;
  color: rgb(101 101 117 / var(--tw-text-opacity));
}

.c2 {
  font-size: 1.125rem;
  line-height: 1.75rem;
  font-weight: 500;
}

.c3 {
  border-radius: 0.125rem;
  --tw-bg-opacity: 1;
  background-color: rgb(243 244 246 / var(--tw-bg-opacity));
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

.c4 {
  margin-bottom: 0.5rem;
  margin-top: 0.75rem;
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
}

.c5 {
  font-size: 1.875rem;
  line-height: 2.25rem;
  font-weight: 700;
  --tw-text-opacity: 1;
  color: rgb(26 25 25 / var(--tw-text-opacity));
}

.c6 {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  --tw-text-opacity: 1;
  color: rgb(255 39 39 / var(--tw-text-opacity));
}

.c7 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 300;
  --tw-text-opacity: 1;
  color: rgb(101 101 117 / var(--tw-text-opacity));
}

.c8 {
  margin-top: 0.125rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 500;
  --tw-text-opacity: 1;
  color: rgb(82 82 86 / var(--tw-text-opacity));
}

<div>
  <div
    class="c0"
    data-testid="statistic-card"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Title
      </div>
      <div
        class="c3"
      >
        Today
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        $9,000
      </div>
      <div
        class="c6"
      >
        <svg
          data-testid="arrow-down"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M862 465.3h-81c-4.6 0-9 2-12.1 5.5L550 723.1V160c0-4.4-3.6-8-8-8h-60c-4.4 0-8 3.6-8 8v563.1L255.1 470.8c-3-3.5-7.4-5.5-12.1-5.5h-81c-6.8 0-10.5 8.1-6 13.2L487.9 861a31.96 31.96 0 0 0 48.3 0L868 478.5c4.5-5.2.8-13.2-6-13.2z"
          />
        </svg>
        <span>
          29
          %
        </span>
      </div>
    </div>
    <div
      class="c7"
    >
      Compared to $
      12000
       yesterday
    </div>
    <div
      class="c8"
    >
      <div>
        Last week Income
      </div>
      <div>
        $50,000
      </div>
    </div>
  </div>
</div>
`;
